web-app/src/main/java/com/google/sps/data/QuadTree.java

```diff
@@ -56,4 +58,85 @@ public void printTree() {
     System.out.printf("%n");
   }
 
+  public void insert(PoliceReport report) {
+    Node leafNode = findLeaf(root, report);
+    leafNode.numReports += 1;
+    leafNode.reports.add(report);
+    // If max capacity has been exceeded, create child nodes
+    if (leafNode.numReports > reportCapacity && leafNode.depth < maxDepth) {
+      leafNode.children = reallocateReports(leafNode.reports, leafNode.coordinates, leafNode.depth);
+      leafNode.leaf = false;
+      leafNode.reports = null;
+    }
+  }
+
+  // Find a leaf to insert the report in
+  private Node findLeaf(Node currentNode, PoliceReport report) {
+    double reportLat = report.getLat();
+    double reportLng = report.getLng();
+
+    while (!currentNode.leaf) {
+      currentNode.numReports += 1;
+      for (Node child : currentNode.children) {
+        if (child.coordinates.inRectangle(reportLat, reportLng)) {
+          currentNode = child;
+          break;
+        }
+      }
+    }
+
+    return currentNode;
+  }
+
+  private Node[] reallocateReports(ArrayList<PoliceReport> reports, Rectangle coordinates, int depth) {
+    Node[] children = new Node[4];
+
+    // Create empty children
+    for (Direction direction : Direction.values()) {
+      Rectangle newCoordinates;
+      ArrayList<PoliceReport> newReports = new ArrayList<PoliceReport>();
+      int newDepth = depth + 1;
+
+      switch (direction) {
+        case NW:
+          newCoordinates = coordinates.getNW();
+          break;
+        case NE:
+          newCoordinates = coordinates.getNE();
+          break;
+        case SE:
+          newCoordinates = coordinates.getSE();
+          break;
+        case SW:
+          newCoordinates = coordinates.getSW();
+          break;
+        default:
+          System.out.println("Unexpected case in switch statement");
+          return children;
+      }
+
+      Node childNode = new Node(newCoordinates, newReports, newDepth);
+      children[direction.ordinal()] = childNode;
+    }
+
+    for (PoliceReport report : reports) {
+      for (Node childNode : children) {
+        if (childNode.coordinates.inRectangle(report.getLat(), report.getLng())) {
+          childNode.reports.add(report);
+          break;
+        }
+      }
+    }
+
+    // Check if number of reports exceeds maximum
+    for (Node childNode : children) {
```

Comment:
I have multiple for loops here which I don't think will impact run time too significantly since number of reports <= 5 and number of children = 4, but I'm wondering if it would be better to try and condense this all into one loop? I split it up because it makes the code a lot simpler (I can make the assumption that every childNode is a leaf when inserting).

Reply:
I prefer what you have here (separate loops).